File: 1636-sort-array-by-increasing-frequency/1636-sort-array-by-increasing-frequency.java
import java.util.*;

class Solution {
    public int[] frequencySort(int[] nums) {
        HashMap<Integer, Integer> hMap = new HashMap<>();
        
        for (int x : nums) {
            hMap.put(x, hMap.getOrDefault(x, 0) + 1);
        }
        
        List<Integer> ans = new ArrayList<>(hMap.keySet());
        ans.sort((x, y) -> {
            int freqComparison = Integer.compare(hMap.get(x), hMap.get(y));
            return freqComparison != 0 ? freqComparison : Integer.compare(y, x);
        });

        int ind = 0;
        int[] res = new int[nums.length];
        
        for (int i = 0; i < ans.size(); ++i) {
            for (int j = 0; j < hMap.get(ans.get(i)); ++j) {
                res[ind++] = ans.get(i);
            }
        }
        
        return res;
    }
}

File: 1636-sort-array-by-increasing-frequency/1636-sort-array-by-increasing-frequency.cpp
class Solution {
public:
    vector<int> frequencySort(vector<int>& nums) {
        unordered_map<int, int> hMap;
        
        for (int x : nums) {
            hMap[x]++;
        }
        
        vector<int> ans;
        for (auto& pair : hMap) {
            ans.push_back(pair.first);
        }
        
        sort(ans.begin(), ans.end(), [&](int x, int y) {
            if (hMap[x] != hMap[y]) {
                return hMap[x] < hMap[y];
            } else {
                return x > y;
            }
        });

        vector<int> res;
        for (int i = 0; i < ans.size(); ++i) {
            for (int j = 0; j < hMap[ans[i]]; ++j) {
                res.push_back(ans[i]);
            }
        }
        
        return res;
    }
};
